Implements CCPP-ized tropopause_find

jimmielin opened the pull request:
Companion PR: https://github.com/ESCOMP/atmospheric_physics/pull/112

Note: I think I brought in more changes than necessary in `ncar_ccpp` due to the my `atmospheric_physics` branch being based off `development` rather than the released used in CAM-SIMA. I can do a rebase if necessary...

## Changes made to support CCPP-ized tropopause_find
* Copied utility module `interpolate_data.F90` from CAM
* Added utility module `tropopause_read_file.F90` to provide climatological tropopause data to tropopause_find from CAM-SIMA
* `&tropopause_nl tropopause_climo_file` is read from `tropopause_read_file.F90` inside SIMA and not in the parameterization, to avoid passing this file name back and forth between the parameterization to CAM-SIMA
* Adds `get_curr_calday()` result to new standard name variable `fractional_calendar_days_on_end_of_current_timestep`
* Submodule update (possibly need a rebase?)

I recommend manually merging in the history PR for testing. It might be easier to merge my branch (`jimmielin:hplin/tropopause_find`) into `peverwhee:history_court`.

jimmielin commented:
Verified that CAM and CAM-SIMA should give the same results now (via good old `write(6,*)`...), although not in the history tape. The same numbers written to the history tape in CAM and CAM-SIMA differ slightly.

```
 TROPP_T   (ncol,time)  t_index =      1     1
       1352     1352  (   787,     1) (   863,     1) (   128,     1) (   128,     1)
                1352   2.241616210937500E+02   1.797986297607422E+02 7.6E-06  1.838663177490234E+02 1.9E-08  1.838663177490234E+02
                1352   2.241616276711713E+02   1.797986345746522E+02          1.838663253775780E+02          1.838663253775780E+02
                1352  (   787,     1) (   863,     1)
          avg abs field values:    2.005496426305827E+02    rms diff: 4.4E-06   avg rel diff(npos):  1.9E-08
                                   2.005496427290007E+02                        avg decimal digits(ndif):  7.9 worst:  7.4
 RMS TROPP_T                          4.3587E-06            NORMALIZED  2.1734E-08

 TROPP_Z   (ncol,time)  t_index =      1     1
       1352     1352  (   390,     1) (  1206,     1) (   384,     1) (   380,     1)
                1352   1.669084179687500E+04   7.539222167968750E+03 9.7E-04  1.662324023437500E+04 2.2E-08  1.647800390625000E+04
                1352   1.669084221174092E+04   7.539222335188982E+03          1.662324119999111E+04          1.647800487060415E+04
                1352  (   390,     1) (  1206,     1)
          avg abs field values:    1.241134300891041E+04    rms diff: 3.0E-04   avg rel diff(npos):  2.2E-08
                                   1.241134300578205E+04                        avg decimal digits(ndif):  7.8 worst:  7.2
 RMS TROPP_Z                          3.0454E-04            NORMALIZED  2.4538E-08

 TROPP_P   (ncol,time)  t_index =      1     1
       1352     1352  (  1206,     1) (   721,     1) (  1206,     1) (    49,     1)
                1352   3.440330078125000E+04   8.685352539062500E+03 1.9E-03  3.440330078125000E+04 2.3E-08  1.684008398437500E+04
                1352   3.440330267933394E+04   8.685352898358373E+03          3.440330267933394E+04          1.684008495238343E+04
                1352  (  1206,     1) (   721,     1)
          avg abs field values:    1.841534202440250E+04    rms diff: 4.8E-04   avg rel diff(npos):  2.3E-08
                                   1.841534203850526E+04                        avg decimal digits(ndif):  7.8 worst:  7.2
 RMS TROPP_P                          4.8473E-04            NORMALIZED  2.6322E-08

```

The identified "tropopause level" (or rather "probability distribution of tropopause level", set to 1 when that level is the tropopause), which is an integer, does not have this problem:
```
 TROPP_PD   (ncol,lev,time)  t_index =      1     1
               43264  (     5,    10,     1) (     1,     1,     1)
               43264   0.100000000000000E+01   0.000000000000000E+00
               43264   0.100000000000000E+01   0.000000000000000E+00
               43264  (     5,    10,     1) (     1,     1,     1)
          avg abs field values:    3.125000000000000E-02
                                   3.125000000000000E-02
```

CAM has output set to `real*8` (via `ndens = 1`) and so is CAM-SIMA, as verified by the netCDF metadata. Can I ask if differences in output are expected between CAM and CAM-SIMA due to the new history infrastructure? @peverwhee 

The differences look something like

CAM:
```
data:

 TROPP_P =
  29800.9219258364, 26680.3490487544, 23092.8788145002, 22414.4083522716,
```

CAM-SIMA:
```
data:

 TROPP_P =
  29800.921875, 26680.349609375, 23092.87890625, 22414.408203125,
```

Debug printouts (directly using `write(6,*)`) show both CAM and CAM-SIMA having same values computed: (log from CAM-SIMA)
```
dec1772.hsn.de.hpc.ucar.edu 0:  twmo: icol, tP, t, pmid           1   29800.921925836381
dec1772.hsn.de.hpc.ucar.edu 0:  twmo: icol, tP, t, pmid           2   26680.349048754382
dec1772.hsn.de.hpc.ucar.edu 0:  twmo: icol, tP, t, pmid           3   23092.878814500178
dec1772.hsn.de.hpc.ucar.edu 0:  twmo: icol, tP, t, pmid           4   22414.408352271617
```

jimmielin commented:
Updated `generate_registry_data.py` to check for valid numeric values before assuming it's a variable.

jimmielin commented:
After the `fillvalue` change I'm getting very small differences in the cold-point-tropopause history output and bit-for-bit in others, still figuring out why...
```
 TROPF_P   (ncol,time)  t_index =      1     1
         23     1352  (  1017,     1) (  1260,     1) (  1017,     1) (  1017,     1)
                1352   9.999999616903162E+35   3.451314683962001E+03 3.8E+28  9.999999616903162E+35 6.5E-10  9.999999616903162E+35
                1352   1.000000000000000E+36   3.451314683962001E+03          1.000000000000000E+36          1.000000000000000E+36
                1352  (  1017,     1) (  1260,     1)
          avg abs field values:    1.701183366780863E+34    rms diff: 5.0E+27   avg rel diff(npos):  6.5E-10
                                   1.701183431952662E+34                        avg decimal digits(ndif):  7.4 worst:  7.4
 RMS TROPF_P                          4.9967E+27            NORMALIZED  2.9372E-07

 TROPF_T   (ncol,time)  t_index =      1     1
         23     1352  (  1017,     1) (   862,     1) (  1017,     1) (  1017,     1)
                1352   9.999999616903162E+35   1.786520363629979E+02 3.8E+28  9.999999616903162E+35 6.5E-10  9.999999616903162E+35
                1352   1.000000000000000E+36   1.786520363629979E+02          1.000000000000000E+36          1.000000000000000E+36
                1352  (  1017,     1) (   862,     1)
          avg abs field values:    1.701183366780863E+34    rms diff: 5.0E+27   avg rel diff(npos):  6.5E-10
                                   1.701183431952662E+34                        avg decimal digits(ndif):  7.4 worst:  7.4
 RMS TROPF_T                          4.9967E+27            NORMALIZED  2.9372E-07

 TROPF_Z   (ncol,time)  t_index =      1     1
         23     1352  (  1017,     1) (  1045,     1) (  1017,     1) (  1017,     1)
                1352   9.999999616903162E+35   9.015423193848925E+03 3.8E+28  9.999999616903162E+35 6.5E-10  9.999999616903162E+35
                1352   1.000000000000000E+36   9.015423193848925E+03          1.000000000000000E+36          1.000000000000000E+36
                1352  (  1017,     1) (  1045,     1)
          avg abs field values:    1.701183366780863E+34    rms diff: 5.0E+27   avg rel diff(npos):  6.5E-10
                                   1.701183431952662E+34                        avg decimal digits(ndif):  7.4 worst:  7.4
 RMS TROPF_Z                          4.9967E+27            NORMALIZED  2.9372E-07
```

jimmielin commented:
I had to specify `_kind_phys` in the `<initial_value>1.e36_kind_phys</initial_value>` to get the correct definition in `physics_state.F90` and bit-for-bit results with CAM/previous version of CAM-SIMA. The regex in `generate_registry_data.py` has also been updated. Thanks!

nusbaume commented on src/utils/interpolate_data.F90:
```diff
@@ -0,0 +1,1230 @@
+module interpolate_data
```
It's unclear to me what the long-term state of things like lat/lon interpolation will be in CAM-SIMA (e.g. we might end up using external libraries like ESMF to do all the work for us).  Thus I decided to skip reviewing this particular Fortran module. 

That being said, I don't want to us to completely forget about it either (it could certainly use some modifications if we do decide to stick with it long-term).  So when you get the chance could you create an issue in CAM-SIMA reminding us to either clean-it-up or replace it at some point?  Thanks!

jimmielin replied:
Thanks, added #307

jimmielin commented on cime_config/namelist_definition_cam.xml:
```diff
@@ -333,6 +333,22 @@
+  <entry id="tropopause_climo_file">
+    <type>char*256</type>
+    <!--<input_pathname>abs</input_pathname>-->
+    <category>tropo</category>
```
Thanks! Just for future reference, do we need to keep the categories consistent with the existing CAM definitions (https://github.com/ESCOMP/CAM/blob/cam_development/bld/namelist_files/namelist_definition.xml#L6117) or we are free to rename the categories for clarity?

nusbaume replied:
Nope, I don't think there is any requirement to maintain the same category names as before, so feel free to rename them as you see fit!

jimmielin commented on cime_config/namelist_definition_cam.xml:
```diff
@@ -333,6 +333,22 @@
+<!-- Tropopause -->
+  <entry id="tropopause_climo_file">
+    <type>char*256</type>
+    <!--<input_pathname>abs</input_pathname>-->
```
Thanks, will do. Wasn't sure if `<input_pathname>` would still be supported in the future (it threw an error for me when I put it in) but it looks like it isn't used anywhere else in SIMA, so I'll remove it. Not sure if there is a replacement mechanism to validate namelist entries that are paths to files...?

nusbaume replied:
I _think_ the different file path types can now be managed by the use of environment variables in the path names themselves (e.g. `${DIN_LOC_ROOT}`), so I don't believe `input_pathname` will be needed anymore.  Of course if we do find a situation where it is required then we can always add it back in then.

jimmielin replied:
Thanks, removed!

jimmielin commented on src/control/cam_comp.F90:
```diff
@@ -224,6 +232,14 @@ subroutine cam_init(caseid, ctitle, model_doi_url,                         &
+      ! Read tropopause climatology
+      call tropopause_read_climo_file()
+
+      ! Mark variables as initialized so they are not read from initial conditions
+      call mark_as_initialized('tropopause_air_pressure_from_climatology_dataset')
+      call mark_as_initialized('tropopause_calendar_days_from_climatology')
+      call mark_as_initialized('filename_of_tropopause_climatology')
```
Thanks, done!

jimmielin commented on src/data/generate_registry_data.py:
```diff
@@ -26,6 +26,7 @@
     sys.path.append(__SPINSCRIPTS)
 # end if
 _ALL_STRINGS_REGEX = re.compile(r'[A-Za-z][A-Za-z_0-9]+')
+_FORTRAN_NUMERIC_REGEX = r'^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?(_kind_phys)?$'
```
Thanks, done!

jimmielin commented on src/data/generate_registry_data.py:
```diff
@@ -940,9 +941,9 @@ def add_variable(self, newvar):
+        # Exclude NULL and nan variables and valid Fortran numeric values that pass the string regex (e.g. 1.e36, -3.2e5)
         for var in all_strings:
-            if var.lower() not in excluded_initializations:
+            if var.lower() not in excluded_initializations and not bool(re.match(_FORTRAN_NUMERIC_REGEX, newvar.initial_value)):
```
Thanks, fixed!

jimmielin commented on src/data/registry.xml:
```diff
@@ -326,6 +327,12 @@
     </variable>
+    <!-- Date and time variables -->
+    <variable local_name="calday"
+              standard_name="fractional_calendar_days_on_end_of_current_timestep"
```
Thanks for bringing this up, good question... This is used to store the result of `time_manager.F90::get_curr_calday()`, and the range is `(1.0, 366.0)`. So it is only reset once we reach the first day of a new year. I prefer `days` because of this but please let me know if there's a better alternate name.

nusbaume replied:
Thanks for the explanation!  I agree that it would be better described as `days` then, so feel free to leave this as-is!

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+
+  use ccpp_kinds,     only: kind_phys
+  use runtime_obj,    only: unset_str
+  use cam_logfile,    only: iulog
+  use cam_abortutils, only: endrun
+  use spmd_utils,     only: masterproc
```
Thanks for the reminder, fixed!

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+    use pio,          only : file_desc_t, var_desc_t, pio_inq_dimid, pio_inq_dimlen, &
+         pio_inq_varid, pio_get_var, pio_closefile, pio_nowrite
+
+    ! character(len=256), intent(in) :: tropopause_climo_file  ! absolute path of climatology file
```
Thanks, removed.

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+
+!> \section arg_table_tropopause_read_file  Argument Table
+!! \htmlinclude tropopause_read_file.html
+  ! days in year for climatological tropopause pressure data
```
Thanks, fixed!

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+  implicit none
+  private
+
+  public :: tropopause_readnl
```
Thanks, updated!

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+module tropopause_read_file
```
Updated to `tropopause_climo_file`. For consistency with the module name, the subroutine names are updated to `tropopause_climo_readnl` and `tropopause_climo_read_file`.

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+    integer :: nlon, nlat, ntimes
+    integer :: start(3)
+    integer :: count(3)
+    integer, parameter :: dates(12) = (/ 116, 214, 316, 415,  516,  615, &
```
Thanks, updated all instances (and added a comment explaining this variable here) to `tropp_slices`.

jimmielin commented on src/physics/utils/tropopause_read_file.meta:
```diff
@@ -0,0 +1,29 @@
+  name  = tropopause_read_file
+  type  = module
+[ tropp_slices ]
+  standard_name = number_of_months_in_year
```
Thanks for the suggestion. I'm a bit on the fence about this here. Since we changed everything in `tropopause_climo_file` to use `tropp_slices = 12` as dimensions I'm concerned that a future change (however unlikely) of the number of time slices used in the tropopause climatology file would need to separate this back into `tropopause_climo_file` if we lift this into `registry.xml` today.

Especially if there's this hardcoded in `tropopause_climo_file`...
```fortran
    ! YMD (01-16, 02-14, ...) corresponding to the time slices of the climatological
    ! tropopause dataset. Will be converted to day-of-year and stored in tropp_days.
    integer, parameter :: dates(tropp_slices) = (/ 116, 214, 316, 415, 516, 615, &
         716, 816, 915, 1016, 1115, 1216 /)
```

To avoid the hidden effect of `number_of_months_in_year` controlling `tropp_slices` here I wonder if we could just rename `tropp_slices`'s standard name to `number_of_time_slices_in_climatology_dataset`. I checked and such a change would not affect `atmospheric_physics`.

Sorry if it's too much thought into a simple variable that might never change! It just makes me slightly uncomfortable that there's some kind of hidden side effect that will be certainly forgotten months later that `number_of_months_in_year` is hardcoded into the # of tropopause slices.

jimmielin replied:
Opened an issue for this since the atmos_phys new directory structure has to be brought in first

https://github.com/ESCOMP/CAM-SIMA/issues/308

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+       open(newunit=unitn, file=trim(nlfile), status='old')
+       call find_group_name(unitn, 'tropopause_nl', status=ierr)
+       if (ierr == 0) then
+          read(unitn, tropopause_nl, iostat=ierr)
```
Thanks, updated!

jimmielin commented on src/physics/utils/tropopause_read_file.F90:
```diff
@@ -0,0 +1,220 @@
+    !-----------------------------------------------------------------------
+    ierr = pio_inq_dimid( pio_id, 'lat', dimid )
+    ierr = pio_inq_dimlen( pio_id, dimid, nlat )
+    allocate( lat(nlat), stat=ierr )
```
Thanks, added!

jimmielin commented:
Thanks @nusbaume for the review. I've addressed most of the concerns. There are a few questions that I wanted to confirm for some suggested changes, could you please take a look? I appreciate your help!

nusbaume reviewed (changes requested):
Super-close!  Just one last code change and a standard-name change.

jimmielin commented on src/physics/utils/tropopause_climo_read.F90:
```diff
@@ -0,0 +1,240 @@
+  real(kind_phys), public, allocatable :: tropp_days(:)
+
+  ! Private module data
+  character(len=256)                   :: tropopause_climo_file = unset_str
```
Done, thank you!

nusbaume reviewed (approved):
Looks great now, thanks @jimmielin!

peverwhee commented on src/physics/utils/tropopause_climo_read.F90:
```diff
@@ -0,0 +1,241 @@
+    ierr = pio_inq_dimid( pio_id, 'lat', dimid )
+    ierr = pio_inq_dimlen( pio_id, dimid, nlat )
+    allocate( lat(nlat), stat=ierr, errmsg=errmsg )
+    if( ierr /= 0 ) then
+       write(iulog,*) 'tropopause_climo_read_file: lat allocation error = ',ierr
+       call endrun('tropopause_climo_read_file: failed to allocate lat, error = ' // errmsg)
+    end if
```
this isn't critical, but for each of these allocation errors, you could use `check_allocate` instead (in cam_abortutils). Might align it more with existing CAM-SIMA code

jimmielin replied:
Thanks @peverwhee! Looks like I can't pass the `errmsg` suggested by @nusbaume into `check_allocate` (it seems to just take the error code `ierr`). Maybe I could open an issue to add a message argument to `check_allocate` and pass in the `errmsg` returned by `allocate` into it in the future?

peverwhee reviewed (approved):
One tiny comment to take or leave! Thanks @jimmielin

jimmielin commented:
Thanks @nusbaume and @peverwhee for the reviews, merging it in! I'll address the standard names and the `check_allocate` issues in separate PRs following this.
